designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,295 @@
+and comment blocks are used to report global information, e.g., stepsize and metric for NUTS-HMC and stepsize for ADVI.
+The current format doesn't allow for fuller reporting of per-iteration computation, e.g., for the NUTS-HMC sampler,
+successive leapfrog steps.  Therefore we need to decouple the outputs from the inference engine from the outputs from the model.
+
+**The inference run**.  When CmdStan is used to do inference,
+it uses the initial set of the comments in the Stan CSV file to record the complete set of configuration options
+and the final block of comments to record timing information.
+Not recorded explicitly are the chain and iteration information, when running multiple chains,
+or timestamp information for processing events.
+
+For example, if we use the CmdStan interface to run the sampler for a single chain
+by default the output is written to a file `output.csv` and if we use the new `num_chains` argument,
+e.g., `num_chains=4` the output is a set of files `output_1.csv`, ..., `output_4.csv`.
+Under this proposal, we would factor the information `output.csv` as follows:
+
+- Structured non-tabular data (nested dictionaries and or/lists) would be output as JSON:
+
+  + `model_metatdata.json` - information on model variable names, types, dimensions, and block 
+  + `sampler_metatdata.json` - information on the configuration of the sampler
+  + `hmc_parameters.json` - information on the stepsize, metric type, and metric.
+
+- Tabular data would be output either as CSV file format or Apache Arrow binary format:
+
+  + `model_sample` - draws from the sampler on the constrained scale
+  + `model_params_unconstrained` - draws on the unconstrained scale (parameter and transformed parameters only)
+  + `log_prob` - the value of `lp__`
+  + `sampler_state` - outputs from the sampler
+  + `sampler_events` - timestamp plus string descriptor
+
+
+## Current implementation
+
+A CSV file is designed to hold a single table's worth of data in plain text where
+each row of the file contains one row of table data, and all rows have the same number of fields.
+The CSV format is not precisely defined.  Common usage allows
+allows the first row of data to be treated as a row of column labels
+(the "header row"), and also allows comment rows which start with a designated comment prefix character.
+Because the initial focus of the project was MCMC sampling, the CSV file allowed for a
+straightforward representation of the posterior sample one row per draw,
+one column per output from the Stan program.
+Over time, the Stan CSV file has come to be an amalgam of information about the inference engine configuration,
+the algorithm state, and the algorithm outputs.
+The formats of the per-inference algorithm outputs are described in the
+[appendix of the CmdStan User's Guide](https://mc-stan.org/docs/2_29/cmdstan-guide/stan_csv.html)
+
+### Output mechanism:  callback writers
```

Comment:
I think this section should also describe what you want to change this to - will the list of data structures be expanded (e.g, directly passing something like an Eigen matrix)? Will function overloading still be used, or will we define a class with more functions like `write_diagnostic`, `write_draws`, etc? (I advocate for the latter, since the current implementation where a string is implicitly a comment and everything else is a draw is difficult to read, and doesn't allow generalization if you want string to mean two different things)

Reply:
we will define a class with more functions like `write_diagnostic` etc.

Reply:
revised.